Lane Follower Systems into dev

Vaibhav-Hariani opened the pull request:
Post-Comp merge of all lane follower code into dev, with atomic commits. Cleaning up systems to make them more legibile and more clear.

Vaibhav-Hariani committed:
added individual follower class definition

Vaibhav-Hariani committed:
added constants and property for binary warp

Vaibhav-Hariani committed:
added window iteration and drawing

Vaibhav-Hariani committed:
full refactoring of individual follower

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+# Lane Follower for a single camera
+# Applies a perspective transform to an image, and isolates lane lines
```
Please create a `component.toml` with this information.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+# Lane Follower for a single camera
+# Applies a perspective transform to an image, and isolates lane lines
+import numpy as np
+import math
+import cv2
```
Imports should be sorted alphabetically and have whitespace between a standard lib imports and 3rd-party improts. They should also come before 3rd-party imports.

Vaibhav-Hariani replied:
Used isort

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+
+
+class IndividualFollower:
+    # Constants for how we process images
+    # These should be parametrized and better defined at a later date: as part of testing
```
```suggestion
```

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+class IndividualFollower:
+    # Constants for how we process images
+    # These should be parametrized and better defined at a later date: as part of testing
+    NWINDOWS = 9
+    SEARCH_WIDTH = 100
+    MINPIXELS = 40
+    LANE_POLY_SIZE = 2
+
+    # Rendering Parameters
+    BOX_COLOR = (0, 255, 0)
+    LANE_COLOR = (255, 0, 0)
+
+    DRAW_THICKNESS = 2
```
I would make a configuration dataclass so that we can easily pass them to a lane-follower. I would also give them the same defaults.

Vaibhav-Hariani replied:
Created a .toml

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+        self._binary_warped = None
+        self.histogram = None
+
+    # Need to determine if these should remain properties
+    # Or if they should be passed as arguments to Plot_Line
+    @property
+    def _binary_warped(self, value: np.ndarray):
+        self._binary_warped = value
+        self.histogram = np.sum(
+            self.binary_warped[self.binary_warped.shape[0] // 2 :, :], axis=0
+        )
```
I'm not sure I follow why you're making this a property? If this is a getter, shouldn't it be a public member?

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+        )
+
+    def plot_line(self):
+
```
```suggestion
```

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+
+    def plot_line(self):
+
+        if self.binary_warped is None:
```
```suggestion
        if not self.binary_warped:
```

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+        if self.binary_warped is None:
+            raise Exception("no binary warp specified")
+
+        # Image to visualize output
+        out_img = (
+            np.dstack(
+                (self._binary_warped, self._binary_warped, self._binary_warped)
+            )
+            * 255
+        )
```
I would like you to see if you could clean up the numpy business with [einops](https://einops.rocks/).

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+            self.binary_warped[self.binary_warped.shape[0] // 2 :, :], axis=0
+        )
+
+    def plot_line(self):
```
This function is very long and should be divided into smaller chunks that are easy to follow.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+        heading = math.atan(slope)
+        result = cv2.addWeighted(out_img, 1, window_img, 0.3, 0)
+        # TODO: Determine what result really is, and annotate efficiently
+        return result, empty_windows, heading, slope
```
It's better to return some kind of result object (maybe a named tuple) so that it's easy to unpack results.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,172 @@
+import cv2
+
+
+class IndividualFollower:
```
```suggestion
class LaneFollower:
```

Vaibhav-Hariani replied:
Not making this change as lane follower is an overarching class that creates 2 individual followers; one for each camera.

jacobkoziej reviewed (changes requested):
Please split the code into smaller functions so that it's easier to follow. The use of [einops](https://einops.rocks/) might also be nice. I would also add type annotations to variables so that we have a better idea of how the data is getting permuted.

Vaibhav-Hariani committed:
created lf_const.toml, a pyproject.toml, and added type hinting

Vaibhav-Hariani committed:
Added result class to simplify individual follower.
 Seriously need to reconsider structure of line_window1 & line_window2:
These lines are only for rendering, but seem to be half running in the loop and half out. This doesn't make a lot of sense? I think they can be fully replaced with the resultant np.polyval

Vaibhav-Hariani committed:
Tried removing some of the numpy vomit

Vaibhav-Hariani committed:
Merge branch 'dev' into user/Vaibhav-Hariani/lane-follower

Vaibhav-Hariani committed:
Added .toml, piping constants into individual folllower from there for better readability

Vaibhav-Hariani committed:
Building indermediary function for code readability

Vaibhav-Hariani committed:
Broke individual follower up into loop & not loop

Vaibhav-Hariani committed:
Added Lane Follower constants to .toml, beginning effort to move this full module to dev

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+
+import cv2
+import numpy as np
+import toml
+from einops import repeat
```
```suggestion
import toml

from einops import repeat
```

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+from einops import repeat
+
+
+class ifResult:
```
Please follow PEP8 for naming. This can also be made into a `dataclass`.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+
+class IndividualFollower:
+
+    def __init__(self, toml_path='Follower_Consts.toml'):
+
+        self.consts = toml.load(toml_path)["Individual_Follower"]
```
I like the vision, but the execution can be better. IMO TOML should be unmarshaled outside the class and passed as a constants struct.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+    def __init__(self, toml_path='Follower_Consts.toml'):
+
+        self.consts = toml.load(toml_path)["Individual_Follower"]
+        # self.fit is set in Plot_Line
+        self.fit = None
+        # These two are set below
+        self._binary_warped = None
+        self._histogram = None
```
I don't think these comments add much value to the reader. I'd also separate public and private members of the class.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+        self._binary_warped = None
+        self._histogram = None
+
+    # Need to determine if these should remain properties
+    # Or if they should be passed as arguments to Plot_Line
+    @property
+    def _binary_warped(self, value: np.ndarray):
```
`@property` should only get used for public members and without arguments so that they just act as a variable access.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+        )
+
+    def get_white_pixels(self):
+        # Returns an output image
+        NWINDOWS = self.consts["NWINDOWS"]
+        SEARCH_WIDTH = self.consts['SEARCH_WIDTH']
```
See, if this was a config class, we wouldn't need these intermediate variables that need to be made. We could also give default values to the config class so that we don't need to always pass it to `__init__()`

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+        lane_base = np.argmax(self.histogram[:])
+
+        # Image to visualize output
+        out_img = repeat(self._binary_warped, 'h w -> h w c', repeat=3) * 255
```
We still have magic numbers. I'd also give more descriptive names to the dimensions like height, width, and channels.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+        # )
+
+        for window in range(NWINDOWS):
+            window_dims = window * window_height
```
If you went through the effort of making `window_height` an `np.int32`, I think you should do the same with `window`, or make them both Python `int`s.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+            win_x_upper = lane_base + SEARCH_WIDTH
+
+            lower_coords = (win_x_lower, win_y_lower)
+            upper_coords = (win_x_upper, win_y_lower)
+            cv2.rectangle(
```
```suggestion
            upper_coords = (win_x_upper, win_y_lower)

            cv2.rectangle(
```

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+                & (nonzero_x >= win_x_upper)
+            ).nonzero_pixels()[0]
+
+            # This should likely be moved into the if statement: leaving for now
```
For stuff like this, turn it into a Python warning so that we don't forget about it.

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+            # This should likely be moved into the if statement: leaving for now
+            lane_inds.append(white_pix_inds)
+            if len(white_pix_inds) > self.consts["MINPIXELS"]:
+                # np.mean will return a float: We need an exact value
+                lane_base = np.int32(np.mean(nonzero_x[white_pix_inds]))
```
```suggestion
                lane_base = np.int32(np.mean(nonzero_x[white_pix_inds]))
```

jacobkoziej commented on components/lane_follower/src/individual_follower.py:
```diff
@@ -0,0 +1,150 @@
+            )
+        return out_img
+
+    def plot_line(self) -> ifResult:
```
Can you search online how other projects might add code to generate plots for certain things? There has to be a cleaner way to do this.

jacobkoziej commented on components/lane_follower/src/lane_follower.py:
Dead code should just not be committed as it's accessible in Git history.

jacobkoziej commented on components/lane_follower/src/pyproject.toml:
I know we haven't discussed this yet, but here's the current schema for a component: https://github.com/CooperUnion/selfdrive/blob/2f991036e81b64c9dfba9d6e9b2461abfb1dfabf/site_scons/site_tools/Component.py#L9

jacobkoziej commented on components/lane_follower/src/pyproject.toml:
```diff
@@ -0,0 +1,8 @@
+  {name = "Vaibhav Hariani"},
+  {name = "Isaiah Rivera"},
+]
+description = "IGVC lane follower module for a single camera"
```
```suggestion
description = "lane follower for a video feed"
```
Something along those lines.

jacobkoziej commented on components/lane_follower/src/Follower_Consts.toml:
It's probably better to just have these as `dataclasses` that have "sane" defaults.
